keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -341,68 +419,134 @@ Contains a unique identifier for the containing cluster.
 
 *   The identifier **must** either:
     *   be a valid [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS label,
-    *   or be composed of two valid [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS labels separated with a dot.
+    *   or be composed of two valid
+        [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS labels separated
+        with a dot.
 *   The identifier **may** be used as a component in MCS DNS.
 *   The identifier **may** be a human readable description of its cluster.
 
 
 ##### Consumers
 
-*   **Must** be able to rely on the identifier existing, unmodified for the entire duration of its membership in a ClusterSet.
-*   **Should** watch the `cluster.clusterset.k8s.io` property to handle potential changes if they live beyond the ClusterSet membership.
-*   **May** rely on the existence of an identifier for clusters that do not belong to a ClusterSet so long as the implementation provides one.
+*   **Must** be able to rely on the identifier existing, unmodified for the
+    entire duration of its membership in a ClusterSet.
+*   **Should** watch the `cluster.clusterset.k8s.io` property to handle
+    potential changes if they live beyond the ClusterSet membership.
+*   **May** rely on the existence of an identifier for clusters that do not
+    belong to a ClusterSet so long as the implementation provides one.
 
 
 ##### Notable scenarios
 
-**Renaming a cluster**: Since a `cluster.clusterset.k8s.io ClusterProperty` must be immutable for the duration of its *membership* in a given ClusterSet, the property contents can be "changed" by unregistering the cluster from the ClusterSet and reregistering it with the new name.
-
-**Reusing cluster names**: Since an `cluster.clusterset.k8s.io ClusterProperty` has no restrictions on whether or not a ClusterProperty can be repeatable, if a cluster unregisters from a ClusterSet it is permitted under this standard to rejoin later with the same `cluster.clusterset.k8s.io ClusterProperty` it had before. Similarly, a *different* cluster could join a ClusterSet with the same `cluster.clusterset.k8s.io ClusterProperty` that had been used by another cluster previously, as long as both do not have membership in the same ClusterSet at the same time. Finally, two or more clusters may have the same `cluster.clusterset.k8s.io ClusterProperty` concurrently (though they **should** not; see "Uniqueness" above) *as long as* they both do not have membership in the same ClusterSet.
+**Renaming a cluster**: Since a `cluster.clusterset.k8s.io ClusterProperty` must
```

Comment:
This feels a little over-opinionated.  It can be changed by deleting and recreating the object.  That may or may not correspond to register/unregister but we have no way of know and no vested interest in it either.

Reply:
I'm not sure here. It's definitely not enforceable, but has benefits. If we forbid changing the cluster ID without unregister/re-register, it means controllers can cache it forever and not worry about changes. This can save a ton of boilerplate effort and painful edge cases. It is a bad idea to change a cluster's identifier while the cluster is in use. If you must do it, take it "out of use" first.

Reply:
If you are a client who has read this and used it in some durable way, and you crash and restart - you HAVE TO recheck the value.  Whether it was mutated or deleted/created, it could hold a different value.  

The main advantage to declaring it immutable is to force finalizers to trigger.  If that's the goal, perhaps we need to actually have an `immutable` field in ClusterProperty, so we can enforce it server-side?